Ultradns rest provider addition

devulapalli8 opened the pull request.

arghya-neustar committed:
Addition of new module named ultradns-rest

arghya-neustar committed:
Restructured ultradns-rest module

arghya-neustar committed:
Completed migration for getNeustarNetworkStatus, getResourceRecordsOfZone, getResourceRecordsOfDNameByType

sathishvc-neustar committed:
A test case `listsAllProvidersWithCredentials` was failing. Fixed it.

The test case `listsAllProvidersWithCredentials` was failing because
ultradns-rest related info was present but not expected in the return
value of `ListProviders.providerAndCredentialsTable()`. Added the
ultradns-rest related info in the expected value.

arghya-neustar committed:
NPPDEVE-1718 Migration of 6 SOAP API to REST

API migrated for this changes are getNeustarNetworkStatus, getAccountsListOfUser, getResourceRecordsOfZone, getResourceRecordsOfDNameByType, createResourceRecord, updateResourceRecord
Added test classes for changes this migrated API
Added apache common-lang dependency for string utilitiues

arghya-neustar committed:
NPPDEVE-1736 Migration of 6 SOAP API to REST

Developed below 6 REST API for the Sprint
1. getDirectionalPoolsOfZone
2. getDirectionalDNSRecordsForHost
3. getDirectionalDNSRecordsForGroup
4. addDirectionalPool
5. getZonesOfAccount
6. deleteResourceRecord

Added new model classes
Added few unit test-cases for the developed API.
Added feign-fttp client dependency to support HTTP PATCH API
Removed unwanted logging

sathishvc-neustar committed:
7 API changes and 5 or 6 test cases

TODO: Break this down into multiple commits

sathishvc-neustar committed:
4 more test cases

arghya-neustar committed:
Added Jacoco plugin in build.gradle

sathishvc-neustar committed:
Refactor: Moved the class Region in UltraDNSRestGeoSupport.java to its own file in model package.

sathishvc-neustar committed:
Removed commented code from UltraDNSRest.java

arghya-neustar committed:
Added not null check while retrieving RRSets from RRSetList
Added exception handling scenarios in ResourceRecord Service class

arghya-neustar committed:
Modified the way to check NetworkStatus

arghya-neustar committed:
Changed the method signature of UltraDNSMockResponse#getMockErrorResponse

arghya-neustar committed:
Modified condition for delete record
Removed one unwanted method

arghya-neustar committed:
Removed one unused Class named 'NameAndType'

sathishvc-neustar committed:
Printed stack trace on the command line when the denominator executable encounters any error.

sathishvc-neustar committed:
In addRecordToRRPool(), enclosed address in double quotes.

sathishvc-neustar committed:
%7B and %7D were not getting converted to { and } characters in @Body.

If the body does not have contain any interpolation, %7B and %7D were not
getting converted to { and } characters. So, provided a work around.

sathishvc-neustar committed:
If data does not exist, handle the exception thrown by getResourceRecordsOfDNameByType() instead of letting the exception bubble up.

sathishvc-neustar committed:
Added a response RR_SET_WITH_NO_RECORDS in UltraDNSMockResponse.java

sathishvc-neustar committed:
Bug fix: Removing recordsFromRdata() as it is not needed anymore.

sathishvc-neustar committed:
Bug Fix: Passing 255 instead of 0 to getResourceRecordsOfDNameByType()

REST API for getResourceRecordsOfDNameByType() does not accept
a value of 0 for RR type to indicate any RR type. The value of
255 indicates any RR type. So, replaced 0 with 255.

sathishvc-neustar committed:
Trivial: Fixed formatting issue - broke down a few long lines into lines with length within 100 character limit.

sathishvc-neustar committed:
Added UltraDNSRestResourceRecordSetApiMockTest.java and added the test case listWhenNoneMatch()

sathishvc-neustar committed:
Added the test case iterateByNameWhenNoneMatch()

sathishvc-neustar committed:
Added the test case getByNameAndTypeWhenAbsent()

sathishvc-neustar committed:
Added the test case getByNameAndTypeWhenPresent()

sathishvc-neustar committed:
Trivial: Modified JavaDoc comment indicating a return value when there is no return value from mockServer().

sathishvc-neustar committed:
Trivial: Fixed formatting issue - broke down a few long lines into lines with length within 100 character limit.

sathishvc-neustar committed:
Added the test case iterateByNameWhenMatch()

sathishvc-neustar committed:
Removed the hasBody() part of the assertion from the test case getByNameAndTypeWhenAbsent()

sathishvc-neustar committed:
Added the test case putFirstACreatesRoundRobinPoolThenAddsRecordToIt()

sathishvc-neustar committed:
Added the test case putFirstAReusesExistingEmptyRoundRobinPool()

sathishvc-neustar committed:
UltraDNSMockResponse.getMockErrorResponse() accepts an int instead of a string.

So, it resulted in some changes whereever it is invoked.

sathishvc-neustar committed:
Renamed the test cases testAddRRLBPoolWithFQDNAsHostName() to createRRPoolInZoneForNameAndType()

sathishvc-neustar committed:
Renamed the test case testAddRRLBPoolWithSubDomainAsHostName() to testAddRRLBPoolWithNonFQDNAsHostName()

sathishvc-neustar committed:
Renamed the test case testAddRRLBPoolWhenPoolAlreadyExists() to createRRPoolInZoneForNameAndTypeWhenAlreadyExists()

sathishvc-neustar committed:
Renamed the test case testGetRegionsByIdAndName() to getRegionsByIdAndName()

sathishvc-neustar committed:
Renamed the test case testDeleteRRPoolWhenPoolExists() to deleteRRPool()

sathishvc-neustar committed:
Renamed the test case testDeleteRRPoolWhenPoolNotFound() to deleteRRPoolWhenPoolNotFound()

sathishvc-neustar committed:
In a test case, a wrong mock response, PUT_NEW_RR_SET_RESPONSE, was being sent. Fixed it.

Fixed it by sending the correct mock response STATUS_SUCCESS.

sathishvc-neustar committed:
Trivial: Indentation issue - added a space before left brace

sathishvc-neustar committed:
Added the test case putSecondAAddsRecordToExistingPool()

sathishvc-neustar committed:
Added the test case putFirstAAAACreatesRoundRobinPoolThenAddsRecordToIt()

sathishvc-neustar committed:
Added the test case putFirstAAAAReusesExistingEmptyRoundRobinPool()

sathishvc-neustar committed:
Added the test case listWhenThereAreMatches()

sathishvc-neustar committed:
Added the test case getByNameAndTypeWhenZoneAbsent()

sathishvc-neustar committed:
Added the test case testUltraDNSRestMetadata() in UltraDNSRestProviderTest

sathishvc-neustar committed:
Added the test case testUltraDNSRestRegistered() in UltraDNSRestProviderTest

sathishvc-neustar committed:
Added the test case testProviderWiresUltraDNSRestZoneApi() in UltraDNSRestProviderTest

sathishvc-neustar committed:
Added the test case testCredentialsRequired() in UltraDNSRestProviderTest

sathishvc-neustar committed:
Added the test case testTwoPartCredentialsRequired() in UltraDNSRestProviderTest

sathishvc-neustar committed:
Added the test case testViaDagger() in UltraDNSRestProviderTest

sathishvc-neustar committed:
Trivial: Removed commented code before addRRLBPool() in UltraDNSRest

sathishvc-neustar committed:
Added the test case retryOnSystemError() in UltraDNSRestTest

sathishvc-neustar committed:
Added the test case recordsInZoneByNameAndTypeInvalidZone() in UltraDNSRestTest

sathishvc-neustar committed:
Added the test case noRetryOnInvalidUser() in UltraDNSRestTest

arghya-neustar committed:
Added new API End Point
Indented code properly in UltraDNSRestAPI.java
Modified unit test-case accordingly
Added one method in RRSetList.java

sathishvc-neustar committed:
Added the test case deleteRecord() in UltraDNSRestTest

sathishvc-neustar committed:
Fixed the logic in deleteByNameAndType() and deletePool() and added the test case deleteAlsoRemovesPool()

* Fixed the logic in deleteByNameAndType() in UltraDNSRestResourceRecordSetApi
* Fixed the logic in deletePool() in UltraDNSRestRoundRobinPoolApi
* Added the test case deleteAlsoRemovesPool() in UltraDNSRestResourceRecordSetApiMockTest

sathishvc-neustar committed:
Modified UltraDNSRestGeoSupport#regions() to include continents as keys and countries as values in return value.

* As per the clarification given by Vitaliy and Kushal, the UltraDNSRestGeoSupport#regions() is modified so that its return value contains the continents as keys and countries as values.
* Added the method UltraDNSRestGeoSupport#regionsAsRegions() to be used in directional pool APIs by Arghya.
* Added the test case for UltraDNSRestGeoSupport#regions()

sathishvc-neustar committed:
Added the test case regionsAsRegions() in UltraDNSRestGeoSupportTest

sathishvc-neustar committed:
Added the test case deleteNonExistentPool() in UltraDNSRestResourceRecordSetApiMockTest

sathishvc-neustar committed:
Added the test case deleteAnAlreadyDeletedPool01() in UltraDNSRestResourceRecordSetApiMockTest

sathishvc-neustar committed:
Added the test case deleteAnAlreadyDeletedPool02() in UltraDNSRestResourceRecordSetApiMockTest

sathishvc-neustar committed:
Bug fix: Continent (Region) name was not added to the values in the return value of regions(). Added it.

arghya-neustar committed:
Added Changes for getDirectionalDNSGroupDetails, addDirectionalPoolRecord API

arghya-neustar committed:
Added test-cases for GeoResourceRecordSetApi

sathishvc-neustar committed:
Removed the method setChildRegions() since it is never used and is not needed.

sathishvc-neustar committed:
Bug fix: Was using Region#getCode() to return effective code. Fixed it by replacing it with getEffectiveCode().

sathishvc-neustar committed:
Bug fix: Region#toString() was returning JSON in which the string keys and string values were not enclosed in double quotes. Fixed it.

sathishvc-neustar committed:
Bug fix: Changed the logic of how two regions are compared.

* Previously, two regions are considered equal only if all the fields id, code, name, type, effectiveCode, parentRegion, childRegions are all equal.
* Changed that logic so that if the fields id, code, name and type are equal then the two regions are considered equal now.

sathishvc-neustar committed:
Added test cases for Region class

Added the following test case in RegionTest.java to test the Region class:
1. testSetNameAndGetName()
2. testSetCodeAndGetCode()
3. testSetTypeAndGetType()
4. testSetIdAndGetId()
5. testGetEffectiveCodesOfChildRegions()
6. testToString()
7. testEquals()

sathishvc-neustar committed:
Added the field effectiveCodeForGeo in Region class as requested by Arghya. Added appropriate test cases too.

sathishvc-neustar committed:
Bug fix: effectiveCodeForGeo was null because constructor in Region was not being called to create objects. Fixed it.

sathishvc-neustar committed:
Bug fix: Returing effectiveCodeForGeo also from Region#toString()

sathishvc-neustar committed:
Removed the method RRSetList#rrSetByNameAndType() since it is not used and is not needed.

sathishvc-neustar committed:
Bug fix: UltraDNSRestException.PATH_NOT_FOUND_TO_PATCH had a wrong value. It should be 170001.

sathishvc-neustar committed:
Added the test case pathNotFoundInPathWhileDeletingResourceRecord() in UltraDNSRestResourceRecordSetApiMockTest

sathishvc-neustar committed:
Added the test case putFirstNsRecordAddsIt() in UltraDNSRestResourceRecordSetApiMockTest

arghya-neustar committed:
Added partial changes for updateDirectionalPoolRecord suggested by Vitaliy

arghya-neustar committed:
Changing deleteResourceRecord API call to deleteDirectionalPoolRecord API call in UltraDNSRestGeoResourceRecordSetApi

arghya-neustar committed:
Modified changes for geo code conversion

arghya-neustar committed:
Added test-cases for UltraDNSRestGeoResourceRecordSetApi in UltraDNSRestGeoResourceRecordSetApiMockTest

  1. putWhenRegionsMatches()
  2. putWhenRegionsDiffer()
  3. putWhenTTLDiffers()
  4. putAsNewDirectionalRecord()

sathishvc-neustar committed:
Bug fix: Modified the logic of UltraDNSRestResourceRecordSetApi#update()

sathishvc-neustar committed:
Added the test case putSecondNsRecordAddsIt() in UltraDNSRestResourceRecordSetApiMockTest

sathishvc-neustar committed:
Added the test case putMethodInvocationSoThatOneNsRecordIsDeleted() in UltraDNSRestResourceRecordSetApiMockTest

arghya-neustar committed:
Modified the logic for UltraDNSRestGeoResourceRecordSetApi#updateDirectionalPoolRecord()

arghya-neustar committed:
Modified the logic for UltraDNSRestGeoResourceRecordSetApi#updateDirectionalPoolRecord()

arghya-neustar committed:
Bug fix : Modified the logic for GroupGeoRecordByNameTypeCustomIterator#hasNext()

arghya-neustar committed:
Added test-cases for UltraDNSRestGeoResourceRecordSetApi in UltraDNSRestGeoResourceRecordSetApiMockTest

1. listWhenPresent()
2. iterateByNameWhenPresent()
3. iterateByNameAndTypeWhenPresent()

arghya-neustar committed:
Modified the access specifiers for UltraDNSRestGeoSupport#getRegionToTerritories()

arghya-neustar committed:
Addition of code changes made as per code review comments for initial merge request.
Below changes were made:

1. Addition of UltraDNSRestPropertyLoader.java
2. Added test class for UltraDNSRestPropertyLoaderTest.java for UltraDNSRestPropertyLoader.java
3. Added application.properties
4. Bug fix : Modified the logic for ZoneList#getZoneNames()
5. Added test class for UltraDNSRestErrorDecoderTest.java for UltraDNSRestErrorDecoder.java
6. Modified constructor of InvalidatableTokenProvider.java to get durationMillis from application.properties
7. Made resource loading property driven in UltraDNSRestProvider.java
8. Added new ResourceType in ResourceTypeToValue enum
9. Created the Record type from ResourceTypeToValue enum in UltraDNSRestProvider.java

arghya-neustar committed:
Removed unwanted loggers

arghya-neustar committed:
Bug fix : Modified the logic for RRSetList#buildRecords() & RRSetList#isDirectionalRecord()
Removed commets

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestGeoResourceRecordSetApi.java:
```diff
@@ -0,0 +1,413 @@
+    if (!supportedTypes.contains(type)) {
+      return Collections.<ResourceRecordSet<?>>emptyList().iterator();
+    }
+    if ("CNAME".equals(type)) {
```
Use the enum

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestGeoResourceRecordSetApi.java:
```diff
@@ -0,0 +1,413 @@
+    List<Map<String, Object>> recordsLeftToCreate = new ArrayList<Map<String, Object>>(rrset.records());
+    Iterator<DirectionalRecord> iterator = recordsByNameTypeAndQualifier(rrset.name(), rrset.type(), group);
+
+    while (iterator.hasNext()) {
```
I think I have already clarified on the mail that you don't need this logic.
you don’t need the logic to determine which record to add, update or delete for an existing owner.
Just call the PUT of REST & that will automatically take care of this.

So, the high-level logic for the implementation should look something like below:
- Call GET end-point for the input owner, zone & type combination
- If any rrset is returned:
  - Call PUT end-point to update (you already have all the information in the input parameter). Just use the input parameter to map it to the REST input for PUT
- If no rrset is returned:
  - Call POST end-point to create (you already have all the information in the input parameter). Just use the input parameter to map it to the REST input for POST

Please also manually run an integration test on the existing SOAP module & the suggested REST module code to verify that the end results are same.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestResourceRecordSetApi.java:
```diff
@@ -0,0 +1,205 @@
+    List<Record> toUpdate = recordsByNameAndType(rrset.name(), rrset.type());
+    List<Map<String, Object>> toCreate = new ArrayList<Map<String, Object>>(rrset.records());
+
+    for (Iterator<Record> shouldUpdate = toUpdate.iterator(); shouldUpdate.hasNext();) {
```
Same as above. you don’t need the logic to determine which record to add, update or delete for an existing owner.
Just call the PUT of REST & that will automatically take care of this.

So, the high-level logic for the implementation should look something like below:
- Call GET end-point for the input owner, zone & type combination
- If any rrset is returned:
  - Call PUT end-point to update (you already have all the information in the input parameter). Just use the input parameter to map it to the REST input for PUT
- If no rrset is returned:
  - Call POST end-point to create (you already have all the information in the input parameter). Just use the input parameter to map it to the REST input for POST

Please also manually run an integration test on the existing SOAP module & the suggested REST module code to verify that the end results are same.

sathishvc replied:
This is existing logic, not something that I wrote from scratch. To stay as close to the original behaviour as possible, we reused (from SOAP) as much as possible. You suggested the logic you mentioned in your comment and email to Arghya when you tried to clarify his doubt. By then, I had completed this as I didn't face any problems with this.

We executed a few integration test cases on the existing SOAP and REST module code and verified the results. We got the expected results. Full fledged integration testing is still pending though. It will be performed once the code review-modifications-merge cycle is done.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestRoundRobinPoolApi.java:
```diff
@@ -0,0 +1,77 @@
+  }
+
+  boolean isPoolType(String type) {
+    return type.equals("A") || type.equals("AAAA");
```
Use the enum

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestTarget.java:
```diff
@@ -0,0 +1,47 @@
+  private final Provider provider;
+  private final InvalidatableTokenProvider lazyToken;
+
+  static final String SOAP_TEMPLATE = "<?xml version=\"1.0\"?>\n";
```
not required in rest

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/NetworkConnection.java:
```diff
@@ -0,0 +1,29 @@
+  @Override
+  public boolean ok() {
+    try {
+      return "GOOD".equals(api.getNeustarNetworkStatus().getMessage().toUpperCase());
```
What if getMessage() returns null? Better to use equalsIgnorecase and not call toUpperCase on getMessage.

arghya-neustar replied:
Can this end-point {ultradns-url}/v2/status return with message as null?

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/model/Region.java:
```diff
@@ -0,0 +1,238 @@
+  }
+
+  public boolean isRegion() {
+    return this.getType().equals("Region");
```
If type is null, this will throw NPE, always use constant .equals to avoid any NPE.

sathishvc replied:
Can there be a region which will have a null value for its "type"? If yes, then instead of creating and raising a new exception like "RegionTypeIsNullException", this NullPointerException will be simpler. Please suggest.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/model/Region.java:
```diff
@@ -0,0 +1,238 @@
+  }
+
+  public boolean isCountry() {
+    return this.getType().equals("Country");
```
Always use constant .equals to avoid any NPE.

sathishvc replied:
Same as above - Can there be a region which will have a null value for its "type"? If yes, then instead of creating and raising a new exception like "RegionTypeIsNullException", this NullPointerException will be simpler. Please suggest.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/model/Region.java:
```diff
@@ -0,0 +1,238 @@
+package denominator.ultradns.model;
```
I am not a big fan of these multiple nested if else blocks & neither do PMD, so please plan to refactor the code in this class in a better way. Use some design pattern to make it more elegant.

sathishvc replied:
Please let me know which if-else blocks you are referring to.

arghya-neustar committed:
Changed Resource Record type from String to Enum as per review comments

arghya-neustar committed:
Removed unwanted codes & logger

arghya-neustar commented on ultradns-rest/src/main/java/denominator/ultradns/model/GeoInfo.java:
```diff
@@ -0,0 +1,25 @@
+public class GeoInfo {
+
+    private String name;
+    private TreeSet<String> codes = new TreeSet<String>();
```
We got the ultra_rest_api-api-XXX.jar from ultradns where all the DTO are declared. It was declared in the same way, so used the same

arghya-neustar commented on ultradns-rest/src/main/java/denominator/ultradns/model/ZoneList.java:
```diff
@@ -0,0 +1,42 @@
+package denominator.ultradns.model;
```
Yes we ran some E2E test scenario & it was successful
REST method responses being casted to these objects by Feign Framework itself

arghya-neustar committed:
Removed unused import

arghya-neustar committed:
1. Modified access modifiers in model classes
2. Incorporated usage of getter & setter instead direct access to member variable

arghya-neustar committed:
Removed unwanted loggers

arghya-neustar committed:
Restructured below model classes:
1. Account.java
2. AccountList.java
3. Zone.java
4. ZoneProperties.java
5. ZoneList.java

ghost commented:
Transaction atomicity problem with the new implementation.
Scenario: During an update, if a customer provides existing rrset for update & rrset with new owner for create, if the POST call fails, there is no way to rollback the PUT call.

FYI: @vpavlyuk

ghost commented:
@sathishvc, @arghya-neustar, please include static code analysis techniques like PMD, findBugs & checkstyle & fix the violations (there are way too many).

Let me know once done, I will resume my code review.

FYI: @vpavlyuk

sathishvc-neustar committed:
Added Checkstyle, PMD and FindBugs to ultradns-rest

sathishvc-neustar committed:
Replaced mistaken use of & with &&

sathishvc-neustar committed:
Enabled XML report and disabled HTML report for FindBugs since only one can be generated for FindBugs.

sathishvc-neustar committed:
Changed the JaCoCo report location to the default.

arghya-neustar committed:
Resolved issues found in PMD Report

arghya-neustar committed:
Resolved few issues found in Findbugs Report

arghya-neustar committed:
Resolved whitespace issues from Checkstyle Report

arghya-neustar committed:
Resolved Imports issues from Checkstyle Report

arghya-neustar committed:
Resolved Naming issues found on Checkstyle Report

arghya-neustar committed:
Resolved Blocks issues found on Checkstyle Report

arghya-neustar committed:
Resolved Checks issues found on Checkstyle Report

arghya-neustar committed:
Renamed file UltraDNSRestRestZoneApiMockTest.java to UltraDNSRestZoneApiMockTest.java

arghya-neustar committed:
Resolved Design issues found on Checkstyle Report

arghya-neustar committed:
Resolved Design issues found on Checkstyle Report

sathishvc-neustar committed:
Added Region#hashCode()

sathishvc-neustar committed:
Used Region#hashCode() to check if two region objects are equal or not.

sathishvc-neustar committed:
Added default in switch clause as per checkstyle recommendations.

sathishvc-neustar committed:
Renamed childRegions to childRegions1 as it masked the instance field.

sathishvc-neustar committed:
equals() is now called on a string literal instead of on a string returned from a method.

sathishvc-neustar committed:
Modified Region#equals() and Region#hashCode() to use Apache EqualsBuilder and HashCodeBuilder respectively.

sathishvc-neustar committed:
Fixed Checkstyle MagicNumberCheck for constants used in hashCode()

sathishvc-neustar committed:
Fixed line length check - split a very long line into smaller lines.

sathishvc-neustar committed:
Fixed MagicNumberCheck warnings in RegionTest.java

arghya-neustar committed:
Restructured the code base
Created new Util classes
Pulled out service methods from model classes and placed them in static util classes
Modified test classes accordingly

arghya-neustar committed:
Fixed checkstyle warnings

arghya-neustar committed:
Modified string comparison while checking network connection

arghya-neustar committed:
Fixed magic number check warnings found in checkstyle report

arghya-neustar committed:
Fixed magic number check warnings found in checkstyle report
Fixed javadoc warnings found in checkstyle report
Fixed line size warnings found in checkstyle report

arghya-neustar committed:
Changed resource type value from string to Enum in test package

sathishvc-neustar committed:
Added @Override to Region#compareTo()

sathishvc-neustar committed:
Moved getRegionHierarchyAsRegions() and getRegionHierarchy() from Region to RegionUtil.

sathishvc-neustar committed:
Trivial: Added a blank new line.

sathishvc-neustar committed:
Removed Region#getEffectiveCodesOfChildRegions().

sathishvc-neustar committed:
Removed an unused import statement in Region.java

sathishvc-neustar committed:
Renamed Region#setParent() to Region#setParentRegion().

sathishvc-neustar committed:
Removed an unused import statement in RegionTest.java

sathishvc-neustar committed:
Prevented RegionUtil from being instantiated.

sathishvc-neustar committed:
Added the description in a javadoc comment.

sathishvc-neustar committed:
Removed UltraDNSRest#getLoadBalancingPoolsByZone()

sathishvc-neustar committed:
Removed the unused string POOL_WITH_NO_RESOURCE_RECORDS

sathishvc-neustar committed:
Removed the unused string UltraDNSMockResponse.TEMP_VAL

sathishvc-neustar committed:
Removed the unused string UltraDNSMockResponse.RR_SET_WITH_TWO_RECORDS

sathishvc-neustar committed:
Added javadoc descriptions in UltraDNSRest

sathishvc-neustar committed:
Fixed the checkstyle warning - unclosed html tag in UltraDNSRestException

sathishvc-neustar committed:
Added description in a javadoc comment in UltraDNSRestProvider

sathishvc-neustar committed:
Fixed a checkstyle warning by splitting a long line in UltraDNSRestTest

arghya-neustar committed:
Removed UltraDNSRestFilter.java

arghya-neustar committed:
Added description in a javadoc comment in GroupGeoRecordByNameTypeCustomIterator

arghya-neustar committed:
Fixed 1 coding & 1 import checkstyle warnings

arghya-neustar committed:
Checkstyle plugin upgrade

Upgraded checkstyle plugin version from 5.7 to 6.5
Removed module RedundantThrows module from checkstyle configuration as it was removed from version 6.2 onwards.

arghya-neustar committed:
Fixed design warnings found in checkstyle report

arghya-neustar commented:
@Kushalraj-Bhandari @vpavlyuk 
Fixed warnings which came from PMD, findBugs & checkstyle report.
You can resume your code review.

FYI: @sathishvc @devulapalli8

sathishvc-neustar committed:
Fixed Checkstyle FileLengthCheck warning in the file UltraDNSRestGeoSupportTest.java

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestZoneApi.java:
```diff
@@ -0,0 +1,133 @@
+  }
+
+  private String getCurrentAccountName() {
+    AccountList accountList = api.getAccountsListOfUser();
```
So, if a user have multiple accounts, this will create the new zone under the last account returned by GET accounts of user. We should have a new optional parameter in Netflix Denominator for the input account, so that user can provide the account under which he wants to create the zone. 
@sdiamond, please let us know your thoughts.
FYI: @vpavlyuk

arghya-neustar replied:
@Kushalraj-Bhandari @vpavlyuk 
If we to pass the account name we have to change the signature of the put method in denominator-ultradns-rest & denominator-cli module, which will have definitely impact on other module. Because method definition is define in interface & the same contract is implementing by all module.
Let us know your thought

FYI @devulapalli8

Kushalraj-Bhandari replied:
Can't you add the accountName field in Zone model, this way you won't have to change the signature. Make sure this field is non-mandatory, so there won't be any impact on other modules...Thoughts?

arghya-neustar replied:
@Kushalraj-Bhandari 
We can do but can we change any class which belongs to other module (Here it is denominator-model & denominator-cli). Could you please check & let me know

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/util/RegionUtil.java:
```diff
@@ -0,0 +1,98 @@
+    // This class is not supposed to be instantiated.
+    private RegionUtil() { }
+
+    public static Map<String, Collection<String>> getRegionHierarchy(Region region) {
```
The logic inside this method & getRegionHierarchyAsRegions seems to be very similar. Please refactor the code in this class to avoid any code duplication.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestGeoSupport.java:
```diff
@@ -0,0 +1,242 @@
+
+  static final int MAX_EFFECTIVE_CODES_PER_API_CALL = 100;
+
+  public UltraDNSRestGeoSupport() { }
```
Why is this default constructor required? Can we remove this?

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestGeoSupport.java:
```diff
@@ -0,0 +1,242 @@
+
+  @Provides
+  @Named("geo")
+  Map<String, Collection<String>> regions(UltraDNSRest api1) {
```
Why is this parameter required, when we already have the same instance variable available?

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestGeoSupport.java:
```diff
@@ -0,0 +1,242 @@
+  Map<Region, Collection<Region>> regionsAsRegions() {
+    Map<Region, Collection<Region>> availableRegions = new TreeMap<Region, Collection<Region>>();
+
+    Collection<Region> topLevelRegions = buildRegionHierarchyByCallingUltraDNSRest(api);
```
Same code as the region() except the map key. No code duplication please.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestGeoSupport.java:
```diff
@@ -0,0 +1,242 @@
+    return availableRegions;
+  }
+
+  private Collection<Region> buildRegionHierarchyByCallingUltraDNSRest(UltraDNSRest api2) {
```
Same as above, is this parameter really required?

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestGeoSupport.java:
```diff
@@ -0,0 +1,242 @@
+        return region;
+      }
+      TreeSet<Region> childRegions = region.getChildRegions();
+      Region region1 = regionGiven(effectiveCode, childRegions);
```
Using digits in variable names like this is usually not preferred, please rename it to use some logical name instead. Take care of this in all the files.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestGeoSupport.java:
```diff
@@ -0,0 +1,242 @@
+import org.apache.commons.lang.StringUtils;
+
+@Module(injects = UltraDNSRestGeoResourceRecordSetApi.Factory.class, complete = false)
+public class UltraDNSRestGeoSupport {
```
So this single class is doing a lot of stuff, better to refactor this class and break into multiple logical classes.

Kushalraj-Bhandari commented on cli/src/test/java/denominator/cli/DenominatorTest.java:
```diff
@@ -89,6 +89,7 @@ public void listsAllProvidersWithCredentials() {
                              "mock       mem:mock                                            false          ",
                              "route53    https://route53.amazonaws.com                       true           accessKey      accessKey secretKey",
                              "route53    https://route53.amazonaws.com                       true           session        accessKey secretKey sessionToken",
+                             "ultradnsrest https://test-restapi.ultradns.com/v2                false          password       username password",
```
don't forget to change it to PROD rest url before releasing.

sdiamond replied:
and make it property driven as that may make it easier.

arghya-neustar replied:
This is a test class under ultradns-cli module. I think Kushal meant a different place in ultadns-rest module to change this url before releasing, where it is already property driven.

Kushalraj-Bhandari commented on ultradns-rest/src/main/resources/application.properties:
```diff
@@ -0,0 +1,7 @@
+feign.read.timeout.millis=600000
+
+# UltraDNS REST properties
+ultradns.rest.url=https://test-restapi.ultradns.com/v2
```
Sorry the URL comment was meant for this place. Change it to PROD URL before final release.

arghya-neustar replied:
Sure, once the code is tested & frezzed, we will change the url from CTE to PROD in application.properties.

FYI @sdiamond

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestGeoSupport.java:
```diff
@@ -0,0 +1,242 @@
+    TreeSet<String> sortedEffectiveCodes = getSortedEffectiveCodes(topLevelRegions);
+    String commaSeparatedCodes = getCommaSeparatedEffectiveCodes(sortedEffectiveCodes);
+
+    response = api2.getAvailableRegions(commaSeparatedCodes);
```
In one of the comment, you mentioned that "api.getAvailableRegions(commaSeparatedCodes)" accepts utmost 100 effective codes. What if the topLevelRegions > 100? Shouldn't they be handled the similar way?

sathishvc replied:
Can the topLevelRegions be > 100. I checked the responses and they don't appear to be. Do you expect that they will be > 100 in the future?

Kushalraj-Bhandari replied:
Who knows, but its always better that our code handles it irrespectively.

arghya-neustar committed:
Moved one utility method named directionalRecordType from service class to util class.

arghya-neustar committed:
Removed unwanted commented lines

arghya-neustar committed:
Functional Changes on Geo Resource Record Update as per reviewer's comments

1. Modified existing service method in Geo Resource Record Service class
2. In the new logic there will be a GET call followed by a PUT or POST to UltraDNS to update the records.
3. Modified UltraDNS API Call
4. Added unit test case also
5. Resolved warnings reported on pmd, checkstyle, findbugs

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/UltraDNSRestGeoSupport.java:
```diff
@@ -0,0 +1,242 @@
+      Collection<String> list = new ArrayList<String>();
+      boolean codeFound = false;
+
+      if (code.equals("A1") || code.equals("A2") || code.equals("A3")) {
```
should be like "string".equals(strObject)

arghya-neustar committed:
Re-Structured iterator classes

arghya-neustar committed:
Re-Structured classes by it's type & placed it under sub-packages
Made modification on access modifiers to support the re structuring

arghya-neustar committed:
Added loggers & few comments in zone service class

arghya-neustar committed:
Refactored RegionUtil Class

arghya-neustar committed:
Handled Data Not Found Exception while listing Geo Records

arghya-neustar committed:
Modfied constant comparison logic

arghya-neustar committed:
Added code comments & logging

arghya-neustar committed:
Refactored UltraDNSRestGeoSupport Class.
Removed code duplication
Modified one utility method in RegionUtil.java
Modified UltraDNSRestGeoResourceRecordSetApi, GroupGeoRecordByNameTypeCustomIterator due to code refactor.
Modified test class.
Added ignore in few test methods as of now, will remove once source modification will be done

arghya-neustar committed:
Added code comments

arghya-neustar committed:
Refactored UltraDNSRestGeoSupport class

Refactored the existing logic
Modified the tests accordingly

arghya-neustar committed:
Added fix for listing Geo Directional records

arghya-neustar committed:
Bug Fix : Added fix on email update while updating zone

arghya-neustar committed:
Modified String comparison

arghya-neustar committed:
Addition of optional account name parameter while creating zone

Added new optional account name property in zone model
Added this new property in CLI-module
Modified the logic in zone service class in ultradns-rest module
Added junit test case

arghya-neustar committed:
Added header param in UltraDNS API Call

arghya-neustar committed:
Modified varrible name

arghya-neustar committed:
Modified comparison

arghya-neustar committed:
Refactored resource record update logic

Modified the existing put method logic
Modified unit test cases accordingly

arghya-neustar committed:
Enhancement : Modified logic Zone service

Modified comparision
Handled exception
Added few code comments

arghya-neustar committed:
Modified test file after code review incorporation

arghya-neustar committed:
Enhancement : String checks & comparison

arghya-neustar committed:
Bug Fix : Handled 'Data Fot Found' exception in getDirectionalDNSRecordsForHost API call

arghya-neustar committed:
Unit Test : Modified logic to test UltraClient header value in Deniminator request

arghya-neustar committed:
Incorporated review comments for additional account name parameter

arghya-neustar committed:
Bug Fix : Class cast exception while applying ttl on SRV GEO records

srikanth-thallapalli committed:
Updated Remove api from PATCH to DELETE

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/service/UltraDNSRestZoneApi.java:
```diff
@@ -0,0 +1,179 @@
+   */
+  @Override
+  public Iterator<Zone> iterator() {
+    final String accountName = getCurrentAccountName();
```
Did you passed this account name for any specific reason? Why didn't you used zone list REST api which don't require account name? That will list all the zones under the user & would make more sense? Let me know in case you have used the account level zones api for any specific reason.

arghya-neustar replied:
Yes, the UltraDNS SOAP module is getting the list of zone for a particular account. We followed the same logic.

arghya-neustar committed:
Enhancement : Account parameter in zone operation

Modified the logic for list zone.
Changed it from get zone of account to get zone of user.
modified the logic in update zone CLI.
Set the account name in zone object in ultradns-rest module.
Modified unit test cases.

arghya-neustar committed:
Enhancement : rData converstion based of record type

srikanth-thallapalli committed:
Refactored PATCH to DELETE api changes

Removed indexToDelete code
Removed body for DELETE call
Updated test cases

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/service/UltraDNSRestResourceRecordSetApi.java:
```diff
@@ -0,0 +1,311 @@
+    try {
+      rrSetList = api.getResourceRecordsOfDNameByType(zoneName, dName, type);
+    } catch (UltraDNSRestException e) {
+      if (e.code() == UltraDNSRestException.DATA_NOT_FOUND ||
```
Throughout the code, you check that if the returned error is not something, you through back the exception. I think its better to create a common method for this and use it everywhere. It can have a parameter for collection of error codes to check & this logic. This way you can kill all those 5-6 lines of repetitive code at numerous places.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/service/UltraDNSRestResourceRecordSetApi.java:
```diff
@@ -0,0 +1,311 @@
+   * @param type
+   * @param record
+   */
+  private void remove(String name, String type, Record record) {
```
record parameter is unused

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/service/UltraDNSRestResourceRecordSetApi.java:
```diff
@@ -0,0 +1,311 @@
+   * @param record
+   */
+  private void remove(String name, String type, Record record) {
+    int intType = lookup(type);
```
No, you don;t need all this complex logic. Just the logic to call deleteResourceRecord is sufficient. You don't need to call GET before this. Also, you don't need any specific handling to check if it is a pool. In terms of DELETE REST API, deleting a pool is same as deleting a record. The URI is exactly the same. Also, make sure the logic present under UltraDNSRestRoundRobinPoolApi#deletePool is not duplicated. Also, remove the method UltraDNSRest#deleteLBPool, and use UltraDNSRest#deleteResourceRecord instead.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/service/integration/UltraDNSRest.java:
```diff
@@ -0,0 +1,323 @@
+   * @param hostName
+   * @return status.
+   */
+  @Headers("Content-Type: application/json-patch+json")
```
Remove @Headers

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/service/integration/UltraDNSRest.java:
```diff
@@ -0,0 +1,323 @@
+   * @param hostName hostname in which the pool is present in.
+   */
+  @RequestLine("DELETE /zones/{zoneName}/rrsets/{typeCode}/{hostName}")
+  void deleteLBPool(@Param("zoneName") String zoneName,
```
This method is not required, use deleteResourceRecord instead.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/service/UltraDNSRestResourceRecordSetApi.java:
```diff
@@ -0,0 +1,311 @@
+   * @param type
+   */
+  @Override
+  public void deleteByNameAndType(String name, String type) {
```
just call UltraDNSRest#deleteResourceRecord. No for loop required

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/service/UltraDNSRestResourceRecordSetApi.java:
```diff
@@ -0,0 +1,311 @@
+  /**
+   * Deletes records for the specified name and type.
+   *
+   * @param name
```
In all the javadocs, you didn't mentioned any parameter description. This field name can be a zone name, it can be record name, and its pretty hard for anyone to interpret what is the field that you are expecting. Please add it everywhere.

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/service/UltraDNSRestResourceRecordSetApi.java:
```diff
@@ -0,0 +1,311 @@
+
+    if (rrSetList == null || rrSetList.getRrSets() == null) {
+      // Create the resource record set.
+      // If the zone does not exist, should it be created? I think it should not be created. Check the SOAP behaviour.
```
Please be certain what your code will do & modify the comment accordingly & clearly. This is an open source project & people can read your comments.

Kushalraj-Bhandari replied:
Also, things like this should go to the method description.

srikanth-thallapalli committed:
Refactored Resource Record API and javadoc updates

Removed deleteLBPool from UltraDNSRest
Updated deleteByNameAndType to use deleteResourceRecord in
UltraDNSRestResourceRecordSetApi
Removed unused remove from UltraDNSRestResourceRecordSetApi
Updated test cases
Updated javadocs parameter descriptions

arghya-neustar committed:
Enhancement : Re-Structured Exception handling flow

arghya-neustar committed:
Javadoc Update

arghya-neustar committed:
E2E Automation Suite Check In

Kushalraj-Bhandari commented on ultradns-rest/src/main/java/denominator/ultradns/service/integration/UltraDNSRest.java:
```diff
@@ -0,0 +1,310 @@
+   * @return status Request status, It can be successful or pending
+   */
+  @RequestLine("DELETE /zones/{zoneName}/rrsets/{rrType}/{hostName}")
+  Status deleteResourceRecord(@Param("zoneName") String zoneName,
```
Same as deleteResourceRecordByNameType, keep one and delete the other...

srikanth-thallapalli committed:
Removed duplicate deleteResourceRecord api method

Removed deleteResourceRecord and retained deleteResourceRecordByNameType
Updated deleteResourceRecord dependencies with deleteResourceRecordByNameType method
Updated testcases to use deleteResourceRecordByNameType method
